specs/lazy-adr/adr-012-fully-decentralized-based-sequencing.md

```diff
@@ -0,0 +1,243 @@
+# ADR 012: Fully Decentralized Based Sequencing
+
+## Changelog
+
+- 2025-04-09: Initial draft
+- 2025-04-09: Added optional UX optimization where full nodes can relay user rollup txs to base layer
+- 2025-04-09: Added rationale for VerifyBatch utility in fully decentralized setup
+- 2025-04-09: Reworded forkchoice rule to use maxHeightDrift instead of time-based maxLatency
+- 2025-04-10: Added Relaying Costs and Fee Compensation via EVM
+
+## Context
+
+Most rollups today rely on centralized or semi-centralized sequencers to form batches of user transactions, despite the availability of base layers (like Celestia) that provide data availability and canonical ordering guarantees. Centralized sequencers introduce liveness and censorship risks, as well as complexity in proposer election, fault tolerance, and bridge security.
+
+Based sequencing eliminates this reliance by having the base layer determine transaction ordering. However, previous implementations still assumed the existence of a proposer to prepare rollup batches.
+
+This ADR proposes a **fully decentralized based sequencing model** in which **every full node acts as its own proposer** by independently:
+- Reading rollup blobs from the base layer
+- Applying a deterministic forkchoice rule
+- Constructing rollup blocks
+- Executing batches to compute state updates
+
+This approach ensures consistency, removes the need for trusted intermediaries, and improves decentralization and resilience.
+
+## Alternative Approaches
+
+### Centralized Sequencer
+- A designated sequencer collects rollup transactions and publishes them to the base layer.
+- Simpler for UX and latency control, but introduces centralization and failure points.
+
+### Leader-Elected Proposer (e.g., BFT committee or rotating proposer)
+- Some nodes are elected to act as proposers for efficiency.
+- Still introduces trust assumptions, coordination complexity, and MEV-related risks.
+
+### Trusted Rollup Light Client Commitments
+- Rollup blocks are committed to L1 (e.g., Ethereum) and verified by a light client.
+- Adds delay and dependency on L1 finality, and often still relies on centralized proposers.
+
+None of these provide the decentralization and self-sovereignty enabled by a fully deterministic, proposerless, based sequencing model.
+
+## Decision
+
+We adopt a fully decentralized based sequencing model where **every full node in the rollup network acts as its own proposer** by deterministically deriving the next batch using only:
```

Comment:
how does this work for ibc and the sdk? what would be put in the header to verify on the counter party chain?

Reply:
every self-proposing full node can relay the headers to counter party chain, but we will have to accompany some form of validity proofs for the counter party chain to verify as they won't be performing the full DA based verfiication and execution. validity proof needs to include da inclusion and state transition, both. need a ADR for this.